Bull in a China Shop C++ Solution (#856)
cpinitiative/usaco-guide: solutions/USACO open16-bull-in-china-shop.mdx

## Implementation

**Time Complexity:** $\mathcal{O}(K^2 \cdot N^6)$

<LanguageSection>

<CPPSection>

```cpp
#include <iostream>
#include <cstdio>
#include <vector>
using namespace std;
bool check(int piece, int x, int y);

int n;
bool pieces[11][8][8];
vector<int> side[11];
int main() 
{
	freopen("bcs.in", "r", stdin);
	freopen("bcs.out", "w", stdout);
	int k;
	cin >> n >> k;
	char c;
	int left, right, top, bottom;
	for (int i = 0; i <= k; i++)
	{
		left = n - 1;
		right = 0;
		top = n - 1;
		bottom = 0;
		for (int j = 0; j < n; j++)
		{
			for (int l = 0; l < n; l++)
			{
				cin >> c;
				pieces[i][j][l] = (c == '#');
				//find the sides of the piece
				if (pieces[i][j][l])
				{
					bottom = max(bottom, j);
					top = min(top, j);
					right = max(right, l);
					left = min(left, l);
				}
			}
		}
		side[i] = {left, right, top, bottom};
	}
	
	//try all the pieces and shifts to find the correct one
	for (int i = 1; i <= k; i++)
	{
		for (int j = i + 1; j <= k; j++)
		{

> 			for (int idx = side[i][3] - n; idx <= side[i][2]; idx++)

Reviewer: ```suggestion
			for (int idx = side[i][3] - n + 1; idx <= side[i][2]; idx++)
```
Is this more accurate?

---
id: usaco-640
source: USACO Bronze 2016 Open
title: Bull in a China Shop
author: Maggie Liu
---

[Official Analysis](http://www.usaco.org/current/data/sol_bcs_bronze_open16.html)

> ## Explanation

Reviewer 2: Do we really need this explanation??? This code seems really similar to the official one... The only difference is that this is in C++ and the official is in Java. I think you should remove this.

Reviewer 3: who asked for your review?

Reviewer 4: > who asked for your review?

@jessechoe10 is on content author team

Author: It is pretty similar, but the official solution is wrong...

Reviewer 5: How is the official solution wrong?

Author: See the warning in the explanation, if it doesn't make sense then I can explain more.

> We can loop through all possible pairs of pieces and shift them to check if they can make the original figurine. Since we cannot shift a piece so any '#' fall outside the $N \times N$ grid, we need to find the sides of the pieces to determine how far to shift them.

Reviewer 6: I would remove this.
```suggestion

```

<Warning>

> The official solution doesn't check if any '#' fall outside the $N \times N$ grid, which is wrong.

Reviewer 6: You are right the original solution seems wrong. Although it is possible that we are missing something about the statement.

> The official solution doesn't check if any '#' fall outside the $N \times N$ grid, which is wrong.

Reviewer 6: ```suggestion
The official solution doesn't check if any '#' fall outside the $N \times N$ grid.
```
Sounds a little bit rude.

</Warning>

> As we read in the pieces, we find the left, right, top and bottom sides of the piece and store this in the $\texttt{side}$ array. When we are shifting, we shift $\texttt{piece[i]}$ $\texttt{idy}$ units to the left and $\texttt{idx}$ units up, and shift $\texttt{piece[j]}$ $\texttt{jdy}$ units to the left and $\texttt{jdx}$ units up. This will send the piece at $(x + \texttt{idx}, y + \texttt{idy})$ to $(x, y)$ and the piece at $(x + \texttt{jdx}, y + \texttt{jdy})$ to $(x, y)$. So, we see that $\texttt{idy}$ must be bounded by $\texttt{right} - n$ and $\texttt{left}$, and $\texttt{idx}$ must be bounded by $\texttt{bottom} - n$ and $\texttt{top}$, with $\texttt{jdy}$ and $\texttt{jdx}$ bounded similarly.

Reviewer 6: ```suggestion
As we iterate the pieces, we find the left, right, top and bottom sides of the piece and store them in the $\texttt{side}$ array. When we are shifting, we shift $\texttt{piece[i]}$ $\texttt{idy}$ units to the left and $\texttt{idx}$ units up, and shift $\texttt{piece[j]}$ $\texttt{jdy}$ units to the left and $\texttt{jdx}$ units up. This will send the piece at $(x + \texttt{idx}, y + \texttt{idy})$ to $(x, y)$ and the piece at $(x + \texttt{jdx}, y + \texttt{jdy})$ to $(x, y)$. So, we see that $\texttt{idy}$ must be bounded by $\texttt{right} - n$ and $\texttt{left}$, and $\texttt{idx}$ must be bounded by $\texttt{bottom} - n$ and $\texttt{top}$, with $\texttt{jdy}$ and $\texttt{jdx}$ bounded similarly. Because of that, every '#' will fall inside the $N \times N$ grid.
```

Once we have shifted the pieces, if there are two '#' in the same place, or if the result of superimposing the pieces is not the same as the original figurine ($\texttt{pieces[0]}$), then this shift will not work.

> 

Reviewer 6: This too. It is already explained in the editorial. Just put something like: We will do the same as the editorial. But in order to avoid shifting '#' pieces outside de N * N grid, we will save the maximum shifts we can do in one direction with every piece.

	{
		for (int j = i + 1; j <= k; j++)
		{
			for (int idx = side[i][3] - n; idx <= side[i][2]; idx++)
			{
				for (int idy = side[i][1] - n; idy <= side[i][0]; idy++)
				{
					for (int jdx = side[j][3] - n; jdx <= side[j][2]; jdx++)
					{
						for (int jdy = side[j][1] - n; jdy <= side[j][0]; jdy++)
						{
							bool good = true;
							for (int x = 0; x < n; x++)
							{
								for (int y = 0; y < n; y++)
								{
									bool ipiece = check(i, x + idx, y + idy);
									bool jpiece = check(j, x + jdx, y + jdy);
									//two '#' are in the same place
									if (ipiece && jpiece)
									{
										good = false;
										break;
									}
									//the result doesn't match the figurine
									if (pieces[0][x][y] != (ipiece || jpiece))
									{
										good = false;
										break;
									}
								}
								if (!good)
								{
									break;
								}
							}
							if (good)
							{
								cout << i << " " << j << endl;
								return 0;
							}
						}
					}
				}
			}
		}
	}
}

//check if a piece is in bounds and is '#'
bool check(int piece, int x, int y)
{
	return x >= 0 && x < n && y >= 0 && y < n && pieces[piece][x][y];

> }

Reviewer 6: The code looks good! Although I prefer don't like to put "{" in a new line :P

Reviewer 6: Nevertheless wait for more reviews, because as I said before maybe we are missing something about the statement or official solution.
